File: src/crawlers/kised.py
"""창업진흥원 (kised.or.kr) 사업공고 크롤러."""

from __future__ import annotations

import logging
import re
from datetime import datetime

from bs4 import BeautifulSoup

from .base import Article, BaseCrawler

logger = logging.getLogger(__name__)

BASE_URL = "https://www.kised.or.kr"
# 사업공고 게시판
LIST_URL = f"{BASE_URL}/menu.es?mid=a10305010000"


class KisedCrawler(BaseCrawler):
    """창업진흥원 사업공고 크롤러."""

    name = "창업진흥원"

    def crawl(self) -> list[Article]:
        resp = self.fetch(LIST_URL)
        if resp is None:
            return []

        resp.encoding = "utf-8"
        soup = BeautifulSoup(resp.text, "html.parser")
        articles: list[Article] = []

        rows = self._find_rows(soup)
        for row in rows:
            article = self._parse_row(row)
            if article:
                articles.append(article)

        logger.info("[%s] %d건 수집 완료", self.name, len(articles))
        return articles

    def _find_rows(self, soup: BeautifulSoup) -> list:
        selectors = [
            "table.boardList tbody tr",
            "table.bbs_list tbody tr",
            "div.board_list table tbody tr",
            "table.tbl_type tbody tr",
            "div.bbs_list_wrap table tbody tr",
            "table tbody tr",
        ]
        for sel in selectors:
            rows = soup.select(sel)
            if rows:
                return rows
        return []

    def _parse_row(self, row) -> Article | None:
        cols = row.select("td")
        if len(cols) < 2:
            return None

        link = row.select_one("a")
        if not link:
            return None

        title = link.get_text(strip=True)
        if not title:
            return None

        url = self._extract_url(link)
        date = self._extract_date(row)

        return Article(title=title, url=url, source=self.name, date=date)

    def _extract_url(self, link) -> str:
        href = link.get("href", "")
        if href and href.startswith("http"):
            return href
        if href and href.startswith("/"):
            return BASE_URL + href

        onclick = link.get("onclick", "")
        if onclick:
            match = re.search(r"'(/[^']+)'", onclick)
            if match:
                return BASE_URL + match.group(1)
            match = re.search(r"(\d{4,})", onclick)
            if match:
                return f"{BASE_URL}/board.es?mid=a10305010000&bid=0005&act=view&otp_id={match.group(1)}"

        return LIST_URL

    def _extract_date(self, element) -> str:
        text = element.get_text()
        match = re.search(r"(\d{4})[.\-/](\d{1,2})[.\-/](\d{1,2})", text)
        if match:
            y, m, d = match.groups()
            try:
                dt = datetime(int(y), int(m), int(d))
                return dt.strftime("%Y-%m-%d")
            except ValueError:
                pass
        return datetime.now().strftime("%Y-%m-%d")

File: src/lib/crawlers/kised.ts
import { BaseCrawler } from './base';
import type { CheerioAPI, AnyNode, Cheerio } from './base';
import type { RawArticle } from './types';

const BASE_URL = 'https://www.kised.or.kr';
const LIST_URL = `${BASE_URL}/menu.es?mid=a10302000000`;

export class KisedCrawler extends BaseCrawler {
  name = '창업진흥원';

  async crawl(): Promise<RawArticle[]> {
    const html = await this.fetch(LIST_URL);
    if (!html) return [];

    const $ = this.parse(html);
    const articles: RawArticle[] = [];

    const items = $('ul.lstyle_list > li').toArray();
    if (items.length === 0) {
      console.warn(`[${this.name}] ul.lstyle_list를 찾을 수 없습니다.`);
      return [];
    }

    for (const item of items) {
      const article = this.parseItem($, item);
      if (article) articles.push(article);
    }

    console.log(`[${this.name}] ${articles.length}건 수집 완료`);
    return articles;
  }

  private parseItem($: CheerioAPI, item: AnyNode): RawArticle | null {
    const $item = $(item);

    const title = $item.find('b.ls_tit').text().trim();
    if (!title) return null;

    let url = LIST_URL;
    const link = $item.find('a[href]').first();
    if (link.length) {
      const href = link.attr('href') ?? '';
      if (href.startsWith('http')) {
        url = href;
      } else if (href.startsWith('/')) {
        url = BASE_URL + href;
      }
    }

    const deadline = this.extractDeadline($, $item);
    const date = new Date().toISOString().slice(0, 10);

    return { title, url, source: this.name, date, deadline };
  }

  private extractDeadline($: CheerioAPI, $item: Cheerio<AnyNode>): string {
    const dts = $item.find('dl dt').toArray();
    const dds = $item.find('dl dd').toArray();

    for (let i = 0; i < Math.min(dts.length, dds.length); i++) {
      const dtText = $(dts[i]).text();
      if (dtText.includes('마감')) {
        const ddText = $(dds[i]).text().trim();
        const match = ddText.match(/(\d{4}-\d{2}-\d{2})/);
        if (match) return match[1];
      }
    }

    for (const dd of dds) {
      const text = $(dd).text().trim();
      const match = text.match(/(\d{4}-\d{2}-\d{2})/);
      if (match) return match[1];
    }

    return '';
  }
}
